cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerRestartService.java

```diff
@@ -0,0 +1,123 @@
+import com.sequenceiq.cloudbreak.dto.StackDtoDelegate;
+import com.sequenceiq.cloudbreak.polling.ExtendedPollingResult;
+import com.sequenceiq.cloudbreak.service.CloudbreakException;
+import com.sequenceiq.cloudbreak.structuredevent.event.CloudbreakEventService;
+
+@Component
+public class ClouderaManagerRestartService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClouderaManagerRestartService.class);
+
+    @Inject
+    private ClouderaManagerPollingServiceProvider clouderaManagerPollingServiceProvider;
+
+    @Inject
+    private PollingResultErrorHandler pollingResultErrorHandler;
+
+    @Inject
+    private CloudbreakEventService eventService;
+
+    @Inject
+    private ClouderaManagerApiFactory clouderaManagerApiFactory;
+
+    public int doRestartServicesIfNeeded(ApiClient apiClient, StackDtoDelegate stack, boolean rollingRestartEnabled) throws ApiException, CloudbreakException {
+        LOGGER.debug("Restarting Cloudera Manager services, rollingRestartEnabled: {}", rollingRestartEnabled);
+        ClustersResourceApi clustersResourceApi = clouderaManagerApiFactory.getClustersResourceApi(apiClient);
+        Optional<ApiCommand> optionalActiveRestartCommand = findActiveRestartCommand(stack, clustersResourceApi, rollingRestartEnabled);
+        if (optionalActiveRestartCommand.isPresent()) {
+            LOGGER.debug("Restart for Cluster services is already running with id: [{}]", optionalActiveRestartCommand.get().getId());
+            return waitForRestartExecution(apiClient, stack, optionalActiveRestartCommand.get());
+        } else {
+            LOGGER.info("Calling restart command. rollingRestartEnabled {}", rollingRestartEnabled);
+            ApiCommand restartCommand = rollingRestartEnabled ?
+                    executeRollingRestartCommand(apiClient, stack, clustersResourceApi) :
+                    executeRestartCommand(stack, clustersResourceApi);
+            eventService.fireCloudbreakEvent(stack.getId(), UPDATE_IN_PROGRESS.name(), CLUSTER_CM_CLUSTER_SERVICES_RESTARTING);
+            return waitForRestartExecution(apiClient, stack, restartCommand);
+        }
+    }
+
+    public void waitForRestartExecutionIfPresent(ApiClient apiClient, StackDtoDelegate stack, boolean rollingRestartEnabled)
+            throws ApiException, CloudbreakException {
+        ClustersResourceApi clustersResourceApi = clouderaManagerApiFactory.getClustersResourceApi(apiClient);
+        Optional<ApiCommand> optionalRestartCommand = findActiveRestartCommand(stack, clustersResourceApi, rollingRestartEnabled);
+        waitForRestartExecution(apiClient, stack, optionalRestartCommand.orElse(null));
+    }
+
+    private int waitForRestartExecution(ApiClient apiClient, StackDtoDelegate stack, ApiCommand restartCommand) throws CloudbreakException {
+        if (Objects.isNull(restartCommand)) {
+            LOGGER.debug("There is no running restart command.");
+            return 0;
+        } else {
```

Comment:
I have the same issue as previously, returning `0` is misleading, as it's not handled with special care. I think we can expect that the restartcommand is not null here, and the caller should ensure that this method is invoked with non null restartCommand.